src/modules/dwm.cpp

```diff
@@ -0,0 +1,493 @@
+      }
+
+      // This event is for keeping track of the currently focused monitor
+      m_ipc->on_monitor_focus_change = [this](const dwmipc::MonitorFocusChangeEvent& ev) {
+        this->on_monitor_focus_change(ev);
+      };
+      m_ipc->subscribe(dwmipc::Event::MONITOR_FOCUS_CHANGE);
+        m_log.info("%s: Subscribed to monitor focus change", name());
+    } catch (const dwmipc::IPCError& err) {
+      throw module_error(err.what());
+    }
+  }
+
+  void dwm_module::stop() {
+    try {
+      m_log.info("%s: Disconnecting from socket", name());
+      m_ipc.reset(nullptr);
+    } catch (...) {
+    }
+
+    event_module::stop();
+  }
+
+  auto dwm_module::has_event() -> bool {
+    try {
+      return m_ipc->handle_event();
+    } catch (const dwmipc::SocketClosedError& err) {
+      m_log.err("%s: Disconnected from socket: %s", name(), err.what());
+      sleep(chrono::duration<double>(1));
+      return reconnect_dwm();
+    } catch (const dwmipc::IPCError& err) {
+      m_log.err("%s: Failed to handle event (reason: %s)", name(), err.what());
+    }
+    return false;
+  }
+
+  auto dwm_module::update() -> bool {
+    // All updates are handled in has_event
+    return true;
+  }
+
+  auto dwm_module::build_cmd(const char* ipc_cmd, const string& arg) -> string {
+    return EVENT_PREFIX + string(ipc_cmd) + "-" + arg;
+  }
+
+  auto dwm_module::build(builder* builder, const string& tag) const -> bool {
+    m_log.info("%s: Building module", name());
+    if (tag == TAG_LABEL_TITLE) {
+      m_log.info("%s: Building title", name());
+      builder->node(m_title_label);
+    } else if (tag == TAG_LABEL_FLOATING) {
+      if (!m_is_floating) {
+        return true;
+      }
+      m_log.info("%s: Building floating label", name());
+      builder->node(m_floating_label);
+    } else if (tag == TAG_LABEL_LAYOUT) {
+      m_log.info("%s: Building layout label", name());
+      if (m_layout_click) {
+        // Toggle between secondary and default layout
+        auto addr = (m_current_layout == m_default_layout ? m_secondary_layout : m_default_layout)->address;
+        builder->cmd(mousebtn::LEFT, build_cmd(CMD_LAYOUT_SET, to_string(addr)));
+        // Set previous layout
+        builder->cmd(mousebtn::RIGHT, build_cmd(CMD_LAYOUT_SET, "0"));
+      }
+      if (m_layout_scroll) {
+        auto addr_next = next_layout(*m_current_layout, m_layout_wrap)->address;
+        auto addr_prev = prev_layout(*m_current_layout, m_layout_wrap)->address;
+        // Set address based on scroll direction
+        auto scroll_down_addr = to_string(m_layout_reverse ? addr_prev : addr_next);
+        auto scroll_up_addr = to_string(m_layout_reverse ? addr_next : addr_prev);
+        builder->cmd(mousebtn::SCROLL_DOWN, build_cmd(CMD_LAYOUT_SET, scroll_down_addr));
+        builder->cmd(mousebtn::SCROLL_UP, build_cmd(CMD_LAYOUT_SET, scroll_up_addr));
+      }
+      builder->node(m_layout_label);
+      if (m_layout_click) {
```

Comment:
This doesn't make a difference here, but in general, command tags should be closed in the reverse order they are opened in. So first closing the scroll tags and then the click tags (they're the same right now, but that might change in the future).

Reply:
Ah, didn't catch that. I'll change it for the sake of consistency.

Reply:
See a9fbf7cd5ce0fa67582c823cc8207c261b92718c